### ISSUE TITLE
[TASK-209] Implement FAQ Generator Agent

### ISSUE THREAD

**Issue by OthmanMohammad** [OthmanMohammad/multi-agent-support-system#97]
# [TASK-209] Implement FAQ Generator Agent

**Task ID:** TASK-209
**Story:** [STORY-002] Knowledge Base Swarm
**Epic:** [EPIC-001] Tier 1: Essential Foundation
**Status:** TODO
**Priority:** P2
**Effort:** 1 day
**Labels:** `task`, `tier-1`, `knowledge-base`, `p2`, `agent`

---

## 📋 Overview

Implement the FAQ Generator agent that automatically generates FAQ entries from frequently asked questions in conversations.

---

## 🎯 Objective

Create an agent that:
1. Finds questions asked >10 times/month
2. Clusters similar questions
3. Generates canonical FAQ entry
4. Includes best answer from conversations
5. Submits for human review

---

## 📂 File Location

```
src/agents/essential/knowledge_base/faq_generator.py
```

---

### Class Structure

```python
from typing import List, Dict
from collections import Counter
from sentence_transformers import SentenceTransformer
import json

from src.agents.base.base_agent import BaseAgent
from src.workflow.state import AgentState

class FAQGenerator(BaseAgent):
    """
    FAQ Generator Agent

    Automatically generates FAQs from conversations.
    """

    def __init__(self):
        super().__init__(
            agent_type="faq_generator",
            model="claude-3-sonnet-20240229",  # Better writing
            temperature=0.3
        )
        self.embedding_model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
        print("✓ FAQ Generator initialized")

    def generate_faqs(
        self,
        days: int = 30,
        min_frequency: int = 10,
        limit: int = 5
    ) -> List[Dict]:
        """
        Generate FAQ entries from conversations

        Args:
            days: Look back period
            min_frequency: Min question frequency
            limit: Max FAQs to generate

        Returns:
            List of FAQ entries
        """
        # Get frequently asked questions
        common_questions = self._find_common_questions(days, min_frequency)

        if not common_questions:
            return []

        # Cluster similar questions
        question_clusters = self._cluster_similar_questions(common_questions)

        # Generate FAQ for each cluster
        faqs = []
        for cluster in question_clusters[:limit]:
            faq = self._generate_faq_entry(cluster)
            faqs.append(faq)

        return faqs

    def _find_common_questions(self, days: int, min_frequency: int) -> List[Dict]:
        """Find commonly asked questions"""
        from datetime import datetime, timedelta
        from src.database.models import Message, Conversation

        cutoff = datetime.utcnow() - timedelta(days=days)

        # Get user messages
        messages = self.db_session.query(Message).join(Conversation).filter(
            Message.role == 'user',
            Conversation.created_at >= cutoff
        ).all()

        # Extract questions (messages ending with ?)
        questions = []
        for msg in messages:
            if '?' in msg.content:
                questions.append({
                    "content": msg.content,
                    "conversation_id": msg.conversation_id
                })

        # Count similar questions
        question_counts = Counter([q["content"].lower() for q in questions])

        # Filter by frequency
        common = [
            {"question": q, "count": count}
            for q, count in question_counts.items()
            if count >= min_frequency
        ]

        return common

    def _cluster_similar_questions(self, questions: List[Dict]) -> List[List[Dict]]:
        """Cluster similar questions together"""
        if not questions:
            return []

        # Generate embeddings
        question_texts = [q["question"] for q in questions]
        embeddings = self.embedding_model.encode(question_texts)

        # Simple clustering (DBSCAN)
        from sklearn.cluster import DBSCAN
        clustering = DBSCAN(eps=0.3, min_samples=2, metric='cosine').fit(embeddings)

        # Group by cluster
        clusters = {}
        for idx, label in enumerate(clustering.labels_):
            if label not in clusters:
                clusters[label] = []
            clusters[label].append(questions[idx])

        return list(clusters.values())

    def _generate_faq_entry(self, question_cluster: List[Dict]) -> Dict:
        """Generate FAQ entry from question cluster"""
        # Get canonical question and answer using LLM
        system_prompt = """You are an FAQ writer.

Given multiple similar questions, create a canonical FAQ entry.

Return JSON:
{
  "question": "Canonical question form",
  "answer": "Clear, concise answer",
  "alternate_phrasings": ["How to...", "Can I..."],
  "category": "billing|technical|usage|api|integrations",
  "confidence": 0.85
}"""

        questions_text = "\n".join([f"- {q['question']}" for q in question_cluster])
        user_prompt = f"""Similar questions:\n{questions_text}\n\nCreate FAQ entry."""

        try:
            response = self.call_llm(system_prompt, user_prompt, max_tokens=512)
            faq = json.loads(response)

            faq["frequency"] = sum(q.get("count", 1) for q in question_cluster)
            faq["status"] = "pending_review"  # Needs human review

            return faq

        except Exception as e:
            print(f"❌ Failed to generate FAQ: {e}")
            return {
                "question": question_cluster[0]["question"],
                "answer": "Answer needed",
                "frequency": len(question_cluster),
                "status": "error"
            }
```

---

## ✅ Acceptance Criteria

- [ ] Finds frequently asked questions
- [ ] Clusters similar questions
- [ ] Generates canonical FAQ form
- [ ] Includes answer
- [ ] Marks for human review
- [ ] Unit tests pass

---

## 🧪 Test Cases

```python
def test_find_common_questions(faq_generator, db_session):
    """Test finding common questions"""
    # Create test messages
    for i in range(15):
        create_message("How do I upgrade?", db_session)

    common = faq_generator._find_common_questions(days=30, min_frequency=10)

    assert len(common) > 0

def test_generate_faq_entry(faq_generator, mock_llm):
    """Test FAQ generation"""
    cluster = [
        {"question": "How do I upgrade?", "count": 10},
        {"question": "Can I upgrade my plan?", "count": 8}
    ]

    faq = faq_generator._generate_faq_entry(cluster)

    assert "question" in faq
    assert "answer" in faq
    assert faq["frequency"] == 18
```

---

**Created:** 2025-11-15
**Assignee:** @OthmanMohammad
